File: heap/502-ipo/IPO.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        projects = sorted(zip(capital, profits))  # sort by required capital
        max_profit_heap = []
        i = 0
        n = len(profits)
        available_capital = w

        for _ in range(k):
            # Add all affordable projects into max heap
            while i < n and projects[i][0] <= available_capital:
                heapq.heappush(max_profit_heap, -projects[i][1])  # push profit as negative
                i += 1

            if not max_profit_heap:  # No project affordable
                break

            available_capital += -heapq.heappop(max_profit_heap)

        return available_capital
        

File: heap/502-ipo/IPO.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        // Min-heap for projects sorted by required capital
        PriorityQueue<int[]> minCapitalHeap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        // Max-heap for profits
        PriorityQueue<int[]> maxProfitHeap = new PriorityQueue<>((a, b) -> b[1] - a[1]);

        for (int i = 0; i < n; i++) {
            minCapitalHeap.offer(new int[]{capital[i], profits[i]});
        }

        int availableCapital = w;
        for (int i = 0; i < k; i++) {
            // Push all affordable projects into maxProfitHeap
            while (!minCapitalHeap.isEmpty() && minCapitalHeap.peek()[0] <= availableCapital) {
                maxProfitHeap.offer(minCapitalHeap.poll());
            }

            if (maxProfitHeap.isEmpty()) break; // No projects can be done

            availableCapital += maxProfitHeap.poll()[1];
        }

        return availableCapital;
    }
}